### PR TITLE
RS: Compose multiple MMA tiles into a single swizzle pattern

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Stacked on https://github.com/NVIDIA/Fuser/pull/2730

This PR is an incremental step towards supporting scheduling multiple tiles of MMA.

The main thing that this PR does is to add a new test `HopperRS.FullSwizzle` and adjust our system to make this test work.

I tried my best to document `HopperRS.FullSwizzle`, so I would recommend starting the review from that test.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
tests are passing now

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

File: tests/cpp/test_mma.cpp [hunk 1]
```diff
@@ -412,6 +412,204 @@ TEST_P(HopperRS, SingleTile) {
...
+  }
+
+  // Create a dummy broadcasting IterDomain to denote that this instruction
+  // is a collective operation over 128 threads. This is a newly created
+  // broadcasting IterDomain and is not connected to other IterDomains in the
+  // TensorDomain. The reason for doing so is because the MMA instruction is
+  // really a collective operation over 128 threads, and by definition there is
+  // no per-thread assignment like "this thread works on this part of the
+  // tensor". It is actually all threads working on all data. For this reason,
+  // the threadIdx.x should not appear anywhere in the index of the tensor.
```
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply tests/cpp/test_mma.cpp:555]: Yes, exactly.

File: tests/cpp/test_mma.cpp [hunk 2]
```diff
@@ -412,6 +412,204 @@ TEST_P(HopperRS, SingleTile) {
...
+  }
+
+  // Inline gmem->register load into the MMA expression at 1.
+  // The shared loop is the loop over multiple tiles.
+  tv0->inlineAt(1);
+  // If TN, then the inner dim is K, then the shared loop is a reduction loop.
+  // This reduction loop does not exist in the register->gmem store, so nothing
+  // to inline.
+  // If TT, then the inner dim is N, then the shared loop is not a reduction
+  // loop. This shared loop exists in the register->gmem store, so we will
```
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply tests/cpp/test_mma.cpp:580]: Sorry, I should rephrase this. I meant the inner dim before reorder.

[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply tests/cpp/test_mma.cpp]: Rephrased as `If TT, then the dim with multiple tiles is N`

File: csrc/device_lower/pass/index.cpp [hunk 3]
```diff
@@ -1624,7 +1624,13 @@ void IndexLowering::handle(const MmaOp* mma) {
     // smem.
     auto tv = mma->inB()->as<TensorView>();
     auto swizzle = getSwizzleMode(tv);
-    auto base_addr = IrBuilder::baseAddressExpr(tv);
+    // Because the entire tile is parallelized on MMA, which are trivial
+    // loops and always have zero loop variables, the result of lowerSrcIndex
```
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply csrc/device_lower/pass/index.cpp:1628]: Just like `ParallelType::Vectorize` and `ParallelType::Bulk`, the `ParallelType::Mma` is not required to be the inner dims. So we do not know which dims to check.

File: csrc/device_lower/pass/allocation.cpp [hunk 4]
```diff
@@ -174,6 +175,74 @@ class AllocationInserter : public kir::ExprMutator {
...
+              id->isDeviceDim()) {
+            continue;
+          }
+          if (ir_utils::isMemoryPartitionedAcross(
+                  info.buffer->getMemoryType(), id->getParallelType())) {
+            continue;
+          }
+          info.allocation_domains->push_back(id);
+          if (gpu_lower->caMap()->idExistsInMap(id, IdMappingMode::LOOP)) {
+            id = gpu_lower->caMap()->getConcreteMappedID(
```
Comment (naoyam) [inline csrc/device_lower/pass/allocation.cpp:227]: Shouldn't this use the loop promotion map?

Comment (zasdfgbnm) [reply csrc/device_lower/pass/allocation.cpp:227]: I believe we should. This part of code is copy-pasted from below without much thinking.😅

Comment (zasdfgbnm) [reply csrc/device_lower/pass/allocation.cpp]: fixed

**Review by naoyam — APPROVED** [review]
LGTM

**Comment by zasdfgbnm** [comment]
!build

**Pull request merged**